---
title: "How Does it Work"
output:
  html_document:
    includes:
      in_header: header.html
      before_body: doc_prefix.html
      after_body: doc_suffix.html
---

Glioblastoma (GBM) is a highly aggressive and incurable form of brain cancer. A large part of GBM malignancy can be attributed to heterogeneity: GBM tumour cells, along with their interactions to the tumour micro-environment, create a complex milieu that ultimately promotes disease progression and causes therapeutic failure.

GBMDeconvoluteR allows users to estimate the relative abundance of various immune and stromal cell populations within bulk GBM expression profiles. Moreover, it also provides abundance estimates for the four neoplastic cell states described be [Neftel et.al (2019)](https://doi.org/10.1016/j.cell.2019.06.024), which are thought to drive glioblastoma malignant cells heterogeneity: neural-progenitor-like (NPC); oligodendrocyte-progenitor-like (OPC); astrocyte-like (AC); and mesenchymal-like (MES).
<br/>
<br/>

<div class="about_header">
  Deconvolution
</div>

Currently, many computational tools/methods exist which allow estimation of cell populations from bulk RNA-sequencing (RNA-seq) data. Broadly speaking, these tools can be conceptually classified into two categories: reference-based (supervised) approaches and reference-free (unsupervised) approaches. Each category has its advantages and limitations, however the reliability of the reference used, is often cited ([Cobos et.al (2020)](https://doi.org/10.1038/s41467-020-19015-1)) as being a major limiting factor when trying to obtain results which have high accuracy. This is because there is often a discrepancy between the biology of the samples and the reference being used to estimate: [Gervin et.al (2019)](https://doi.org/10.1186/s13148-019-0717-y) have shown that samples with different phenotypes to that of the population of interest reduce the performance of reference-based methods.

GBMDeconvoluteR uses a reference-based deconvolution method called [MCPCounter](https://doi.org/10.1186/s13059-016-1070-5) which has been shown by [Sturum et.el (2019)](https://doi.org/10.1093/bioinformatics/btz363) to perform favourably when compared with other methods. To account for the reference reliability issues mentioned above, we derived a set of GBM-specific immune cell markers using 5 publicly available single cell RNA-seq datasets. More information on this process can be found here in our [accompanying paper](https://doi.org/10.1093/neuonc/noad021).
