{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import torch"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GeneName</th>\n",
       "      <th>JAN_P</th>\n",
       "      <th>AAB_P</th>\n",
       "      <th>AAC_P</th>\n",
       "      <th>AAF_P</th>\n",
       "      <th>AAG_P</th>\n",
       "      <th>AAJ_P</th>\n",
       "      <th>AAL_P</th>\n",
       "      <th>AAM_P</th>\n",
       "      <th>AAN_P</th>\n",
       "      <th>...</th>\n",
       "      <th>GLSS-SM-R080_R</th>\n",
       "      <th>GLSS-SM-R080_P</th>\n",
       "      <th>GLSS-SM-R083_R</th>\n",
       "      <th>GLSS-SM-R083_P</th>\n",
       "      <th>GLSS-SM-R088_R</th>\n",
       "      <th>GLSS-SM-R088_P</th>\n",
       "      <th>GLSS-SM-R093_R</th>\n",
       "      <th>GLSS-SM-R093_P</th>\n",
       "      <th>TCGA-06-0125_R</th>\n",
       "      <th>TCGA-06-0125_P</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AAAS</td>\n",
       "      <td>543.0</td>\n",
       "      <td>704.0</td>\n",
       "      <td>471.0</td>\n",
       "      <td>517.0</td>\n",
       "      <td>655.0</td>\n",
       "      <td>431.0</td>\n",
       "      <td>654.0</td>\n",
       "      <td>567.0</td>\n",
       "      <td>681.0</td>\n",
       "      <td>...</td>\n",
       "      <td>2147.0</td>\n",
       "      <td>2187.0</td>\n",
       "      <td>232.0</td>\n",
       "      <td>500.0</td>\n",
       "      <td>2252.0</td>\n",
       "      <td>2696.0</td>\n",
       "      <td>1587.0</td>\n",
       "      <td>2020.0</td>\n",
       "      <td>1084.0</td>\n",
       "      <td>2602.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AADAT</td>\n",
       "      <td>160.0</td>\n",
       "      <td>203.0</td>\n",
       "      <td>117.0</td>\n",
       "      <td>220.0</td>\n",
       "      <td>259.0</td>\n",
       "      <td>266.0</td>\n",
       "      <td>146.0</td>\n",
       "      <td>145.0</td>\n",
       "      <td>107.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1929.0</td>\n",
       "      <td>1312.0</td>\n",
       "      <td>808.0</td>\n",
       "      <td>842.0</td>\n",
       "      <td>325.0</td>\n",
       "      <td>270.0</td>\n",
       "      <td>511.0</td>\n",
       "      <td>782.0</td>\n",
       "      <td>275.0</td>\n",
       "      <td>918.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AAGAB</td>\n",
       "      <td>518.0</td>\n",
       "      <td>785.0</td>\n",
       "      <td>431.0</td>\n",
       "      <td>289.0</td>\n",
       "      <td>639.0</td>\n",
       "      <td>391.0</td>\n",
       "      <td>615.0</td>\n",
       "      <td>460.0</td>\n",
       "      <td>522.0</td>\n",
       "      <td>...</td>\n",
       "      <td>2321.0</td>\n",
       "      <td>1472.0</td>\n",
       "      <td>1362.0</td>\n",
       "      <td>2076.0</td>\n",
       "      <td>1059.0</td>\n",
       "      <td>1085.0</td>\n",
       "      <td>1655.0</td>\n",
       "      <td>2248.0</td>\n",
       "      <td>1569.0</td>\n",
       "      <td>2147.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AAK1</td>\n",
       "      <td>4397.0</td>\n",
       "      <td>4871.0</td>\n",
       "      <td>3418.0</td>\n",
       "      <td>3521.0</td>\n",
       "      <td>5450.0</td>\n",
       "      <td>4107.0</td>\n",
       "      <td>4469.0</td>\n",
       "      <td>3094.0</td>\n",
       "      <td>4695.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1517.0</td>\n",
       "      <td>621.0</td>\n",
       "      <td>252.0</td>\n",
       "      <td>214.0</td>\n",
       "      <td>158.0</td>\n",
       "      <td>273.0</td>\n",
       "      <td>651.0</td>\n",
       "      <td>644.0</td>\n",
       "      <td>812.0</td>\n",
       "      <td>427.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AAR2</td>\n",
       "      <td>458.0</td>\n",
       "      <td>299.0</td>\n",
       "      <td>421.0</td>\n",
       "      <td>206.0</td>\n",
       "      <td>305.0</td>\n",
       "      <td>333.0</td>\n",
       "      <td>328.0</td>\n",
       "      <td>300.0</td>\n",
       "      <td>309.0</td>\n",
       "      <td>...</td>\n",
       "      <td>2120.0</td>\n",
       "      <td>1912.0</td>\n",
       "      <td>195.0</td>\n",
       "      <td>109.0</td>\n",
       "      <td>1137.0</td>\n",
       "      <td>705.0</td>\n",
       "      <td>1571.0</td>\n",
       "      <td>1673.0</td>\n",
       "      <td>1535.0</td>\n",
       "      <td>3084.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5824</th>\n",
       "      <td>ZW10</td>\n",
       "      <td>378.0</td>\n",
       "      <td>270.0</td>\n",
       "      <td>334.0</td>\n",
       "      <td>262.0</td>\n",
       "      <td>318.0</td>\n",
       "      <td>279.0</td>\n",
       "      <td>264.0</td>\n",
       "      <td>235.0</td>\n",
       "      <td>264.0</td>\n",
       "      <td>...</td>\n",
       "      <td>776.0</td>\n",
       "      <td>738.0</td>\n",
       "      <td>817.0</td>\n",
       "      <td>604.0</td>\n",
       "      <td>542.0</td>\n",
       "      <td>664.0</td>\n",
       "      <td>539.0</td>\n",
       "      <td>606.0</td>\n",
       "      <td>744.0</td>\n",
       "      <td>1027.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5825</th>\n",
       "      <td>ZWILCH</td>\n",
       "      <td>75.0</td>\n",
       "      <td>185.0</td>\n",
       "      <td>102.0</td>\n",
       "      <td>62.0</td>\n",
       "      <td>133.0</td>\n",
       "      <td>101.0</td>\n",
       "      <td>151.0</td>\n",
       "      <td>116.0</td>\n",
       "      <td>44.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1344.0</td>\n",
       "      <td>1131.0</td>\n",
       "      <td>2764.0</td>\n",
       "      <td>2753.0</td>\n",
       "      <td>377.0</td>\n",
       "      <td>673.0</td>\n",
       "      <td>563.0</td>\n",
       "      <td>741.0</td>\n",
       "      <td>328.0</td>\n",
       "      <td>1197.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5826</th>\n",
       "      <td>ZWINT</td>\n",
       "      <td>237.0</td>\n",
       "      <td>268.0</td>\n",
       "      <td>324.0</td>\n",
       "      <td>177.0</td>\n",
       "      <td>167.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>200.0</td>\n",
       "      <td>194.0</td>\n",
       "      <td>97.0</td>\n",
       "      <td>...</td>\n",
       "      <td>2900.0</td>\n",
       "      <td>1296.0</td>\n",
       "      <td>2004.0</td>\n",
       "      <td>971.0</td>\n",
       "      <td>1386.0</td>\n",
       "      <td>1799.0</td>\n",
       "      <td>1147.0</td>\n",
       "      <td>1136.0</td>\n",
       "      <td>180.0</td>\n",
       "      <td>1092.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5827</th>\n",
       "      <td>ZXDC</td>\n",
       "      <td>2443.0</td>\n",
       "      <td>1981.0</td>\n",
       "      <td>2225.0</td>\n",
       "      <td>1100.0</td>\n",
       "      <td>925.0</td>\n",
       "      <td>1171.0</td>\n",
       "      <td>1685.0</td>\n",
       "      <td>1076.0</td>\n",
       "      <td>1585.0</td>\n",
       "      <td>...</td>\n",
       "      <td>943.0</td>\n",
       "      <td>1237.0</td>\n",
       "      <td>120.0</td>\n",
       "      <td>212.0</td>\n",
       "      <td>607.0</td>\n",
       "      <td>630.0</td>\n",
       "      <td>543.0</td>\n",
       "      <td>1164.0</td>\n",
       "      <td>1053.0</td>\n",
       "      <td>2158.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5828</th>\n",
       "      <td>ZYX</td>\n",
       "      <td>3112.0</td>\n",
       "      <td>4670.0</td>\n",
       "      <td>2421.0</td>\n",
       "      <td>2649.0</td>\n",
       "      <td>3495.0</td>\n",
       "      <td>1726.0</td>\n",
       "      <td>5512.0</td>\n",
       "      <td>3724.0</td>\n",
       "      <td>4607.0</td>\n",
       "      <td>...</td>\n",
       "      <td>5036.0</td>\n",
       "      <td>2626.0</td>\n",
       "      <td>22.0</td>\n",
       "      <td>127.0</td>\n",
       "      <td>6316.0</td>\n",
       "      <td>8508.0</td>\n",
       "      <td>1919.0</td>\n",
       "      <td>3447.0</td>\n",
       "      <td>11983.0</td>\n",
       "      <td>35720.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5829 rows × 609 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     GeneName   JAN_P   AAB_P   AAC_P   AAF_P   AAG_P   AAJ_P   AAL_P   AAM_P  \\\n",
       "0        AAAS   543.0   704.0   471.0   517.0   655.0   431.0   654.0   567.0   \n",
       "1       AADAT   160.0   203.0   117.0   220.0   259.0   266.0   146.0   145.0   \n",
       "2       AAGAB   518.0   785.0   431.0   289.0   639.0   391.0   615.0   460.0   \n",
       "3        AAK1  4397.0  4871.0  3418.0  3521.0  5450.0  4107.0  4469.0  3094.0   \n",
       "4        AAR2   458.0   299.0   421.0   206.0   305.0   333.0   328.0   300.0   \n",
       "...       ...     ...     ...     ...     ...     ...     ...     ...     ...   \n",
       "5824     ZW10   378.0   270.0   334.0   262.0   318.0   279.0   264.0   235.0   \n",
       "5825   ZWILCH    75.0   185.0   102.0    62.0   133.0   101.0   151.0   116.0   \n",
       "5826    ZWINT   237.0   268.0   324.0   177.0   167.0   100.0   200.0   194.0   \n",
       "5827     ZXDC  2443.0  1981.0  2225.0  1100.0   925.0  1171.0  1685.0  1076.0   \n",
       "5828      ZYX  3112.0  4670.0  2421.0  2649.0  3495.0  1726.0  5512.0  3724.0   \n",
       "\n",
       "       AAN_P  ...  GLSS-SM-R080_R  GLSS-SM-R080_P  GLSS-SM-R083_R  \\\n",
       "0      681.0  ...          2147.0          2187.0           232.0   \n",
       "1      107.0  ...          1929.0          1312.0           808.0   \n",
       "2      522.0  ...          2321.0          1472.0          1362.0   \n",
       "3     4695.0  ...          1517.0           621.0           252.0   \n",
       "4      309.0  ...          2120.0          1912.0           195.0   \n",
       "...      ...  ...             ...             ...             ...   \n",
       "5824   264.0  ...           776.0           738.0           817.0   \n",
       "5825    44.0  ...          1344.0          1131.0          2764.0   \n",
       "5826    97.0  ...          2900.0          1296.0          2004.0   \n",
       "5827  1585.0  ...           943.0          1237.0           120.0   \n",
       "5828  4607.0  ...          5036.0          2626.0            22.0   \n",
       "\n",
       "      GLSS-SM-R083_P  GLSS-SM-R088_R  GLSS-SM-R088_P  GLSS-SM-R093_R  \\\n",
       "0              500.0          2252.0          2696.0          1587.0   \n",
       "1              842.0           325.0           270.0           511.0   \n",
       "2             2076.0          1059.0          1085.0          1655.0   \n",
       "3              214.0           158.0           273.0           651.0   \n",
       "4              109.0          1137.0           705.0          1571.0   \n",
       "...              ...             ...             ...             ...   \n",
       "5824           604.0           542.0           664.0           539.0   \n",
       "5825          2753.0           377.0           673.0           563.0   \n",
       "5826           971.0          1386.0          1799.0          1147.0   \n",
       "5827           212.0           607.0           630.0           543.0   \n",
       "5828           127.0          6316.0          8508.0          1919.0   \n",
       "\n",
       "      GLSS-SM-R093_P  TCGA-06-0125_R  TCGA-06-0125_P  \n",
       "0             2020.0          1084.0          2602.0  \n",
       "1              782.0           275.0           918.0  \n",
       "2             2248.0          1569.0          2147.0  \n",
       "3              644.0           812.0           427.0  \n",
       "4             1673.0          1535.0          3084.0  \n",
       "...              ...             ...             ...  \n",
       "5824           606.0           744.0          1027.0  \n",
       "5825           741.0           328.0          1197.0  \n",
       "5826          1136.0           180.0          1092.0  \n",
       "5827          1164.0          1053.0          2158.0  \n",
       "5828          3447.0         11983.0         35720.0  \n",
       "\n",
       "[5829 rows x 609 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def pyLoadData(ext, path):\n",
    "    try:\n",
    "        # Switch-like structure to read the file based on its extension\n",
    "        if ext == 'csv':\n",
    "            data = pd.read_csv(path)\n",
    "        elif ext == 'tsv':\n",
    "            data = pd.read_csv(path, sep='\\t')\n",
    "        elif ext == 'xlsx':\n",
    "            data = pd.read_excel(path, engine='openpyxl')\n",
    "        else:\n",
    "            raise ValueError(\"Invalid file; Please upload a .csv, .tsv, or .xlsx file\")\n",
    "    except Exception as e:\n",
    "        raise ValueError(f\"Failed to read the file: {e}\")\n",
    "    \n",
    "    return data\n",
    "\n",
    "data = pyLoadData(\"csv\", \"../data/GBMPurity-test-input.csv\")\n",
    "data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "def tpm(X: np.ndarray, lengths: np.ndarray):\n",
    "    \"\"\"\n",
    "    Calculate TPM (Transcripts Per Million) normalization for RNA-seq data.\n",
    "\n",
    "    Parameters:\n",
    "    X (np.ndarray): 2D array of raw read counts (sample x genes).\n",
    "    lengths (np.ndarray): 1D array of feature lengths (e.g., gene lengths).\n",
    "\n",
    "    Returns:\n",
    "    np.ndarray: TPM normalized values.\n",
    "    \"\"\"\n",
    "    \n",
    "    if X.shape[1] != lengths.shape[0]:\n",
    "        raise ValueError(\"The number of rows in X must match the length of lengths\")\n",
    "    \n",
    "    # Calculate RPK (Reads Per Kilobase)\n",
    "    rpk = np.divide(X, lengths)\n",
    "    \n",
    "    # Calculate the scaling factor\n",
    "    scaling_factor = np.nansum(rpk, axis=1).reshape(-1, 1)\n",
    "    \n",
    "    # Calculate TPM\n",
    "    tpm = (rpk / scaling_factor) * 1e4\n",
    "    \n",
    "    return tpm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "feature_name      AAAS  AADAT   AAGAB    AAK1    AAR2  AARSD1  AASDHPPT  \\\n",
      "JAN_P            543.0  160.0   518.0  4397.0   458.0   429.0     706.0   \n",
      "AAB_P            704.0  203.0   785.0  4871.0   299.0   397.0    1078.0   \n",
      "AAC_P            471.0  117.0   431.0  3418.0   421.0   360.0     788.0   \n",
      "AAF_P            517.0  220.0   289.0  3521.0   206.0   326.0     555.0   \n",
      "AAG_P            655.0  259.0   639.0  5450.0   305.0   320.0     747.0   \n",
      "...                ...    ...     ...     ...     ...     ...       ...   \n",
      "GLSS-SM-R088_P  2696.0  270.0  1085.0   273.0   705.0   680.0    1116.0   \n",
      "GLSS-SM-R093_R  1587.0  511.0  1655.0   651.0  1571.0  1209.0    2911.0   \n",
      "GLSS-SM-R093_P  2020.0  782.0  2248.0   644.0  1673.0  1132.0    2502.0   \n",
      "TCGA-06-0125_R  1084.0  275.0  1569.0   812.0  1535.0   859.0    2860.0   \n",
      "TCGA-06-0125_P  2602.0  918.0  2147.0   427.0  3084.0  1477.0    3777.0   \n",
      "\n",
      "feature_name      AASS   AATK   ABCA5  ...  ZSCAN21  ZSCAN5A  ZSCAN9  ZSWIM4  \\\n",
      "JAN_P           1541.0  717.0  1540.0  ...    380.0    389.0   381.0   524.0   \n",
      "AAB_P           3087.0  284.0   331.0  ...    384.0    394.0   403.0   306.0   \n",
      "AAC_P           1421.0   68.0   510.0  ...    351.0    220.0   244.0   239.0   \n",
      "AAF_P           1837.0  104.0  1460.0  ...    368.0    285.0   387.0   476.0   \n",
      "AAG_P            874.0  126.0   526.0  ...    159.0     97.0   331.0   263.0   \n",
      "...                ...    ...     ...  ...      ...      ...     ...     ...   \n",
      "GLSS-SM-R088_P  1273.0  977.0   558.0  ...    517.0    127.0   318.0   457.0   \n",
      "GLSS-SM-R093_R   569.0  456.0  1004.0  ...    360.0    201.0   451.0   236.0   \n",
      "GLSS-SM-R093_P  1099.0   61.0  1286.0  ...    593.0    371.0   830.0   236.0   \n",
      "TCGA-06-0125_R   900.0  723.0   870.0  ...    611.0    332.0   404.0   511.0   \n",
      "TCGA-06-0125_P  1743.0  454.0  1431.0  ...   1037.0    511.0  1197.0   878.0   \n",
      "\n",
      "feature_name    ZSWIM6    ZW10  ZWILCH   ZWINT    ZXDC      ZYX  \n",
      "JAN_P           1222.0   378.0    75.0   237.0  2443.0   3112.0  \n",
      "AAB_P           1102.0   270.0   185.0   268.0  1981.0   4670.0  \n",
      "AAC_P           1255.0   334.0   102.0   324.0  2225.0   2421.0  \n",
      "AAF_P            886.0   262.0    62.0   177.0  1100.0   2649.0  \n",
      "AAG_P           1164.0   318.0   133.0   167.0   925.0   3495.0  \n",
      "...                ...     ...     ...     ...     ...      ...  \n",
      "GLSS-SM-R088_P   799.0   664.0   673.0  1799.0   630.0   8508.0  \n",
      "GLSS-SM-R093_R  1201.0   539.0   563.0  1147.0   543.0   1919.0  \n",
      "GLSS-SM-R093_P  1433.0   606.0   741.0  1136.0  1164.0   3447.0  \n",
      "TCGA-06-0125_R  1272.0   744.0   328.0   180.0  1053.0  11983.0  \n",
      "TCGA-06-0125_P  1832.0  1027.0  1197.0  1092.0  2158.0  35720.0  \n",
      "\n",
      "[608 rows x 5829 columns]\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>feature_name</th>\n",
       "      <th>AAAS</th>\n",
       "      <th>AADAT</th>\n",
       "      <th>AAGAB</th>\n",
       "      <th>AAK1</th>\n",
       "      <th>AAR2</th>\n",
       "      <th>AARSD1</th>\n",
       "      <th>AASDHPPT</th>\n",
       "      <th>AASS</th>\n",
       "      <th>AATK</th>\n",
       "      <th>ABCA5</th>\n",
       "      <th>...</th>\n",
       "      <th>ZSCAN21</th>\n",
       "      <th>ZSCAN5A</th>\n",
       "      <th>ZSCAN9</th>\n",
       "      <th>ZSWIM4</th>\n",
       "      <th>ZSWIM6</th>\n",
       "      <th>ZW10</th>\n",
       "      <th>ZWILCH</th>\n",
       "      <th>ZWINT</th>\n",
       "      <th>ZXDC</th>\n",
       "      <th>ZYX</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>JAN_P</th>\n",
       "      <td>543.0</td>\n",
       "      <td>160.0</td>\n",
       "      <td>518.0</td>\n",
       "      <td>4397.0</td>\n",
       "      <td>458.0</td>\n",
       "      <td>429.0</td>\n",
       "      <td>706.0</td>\n",
       "      <td>1541.0</td>\n",
       "      <td>717.0</td>\n",
       "      <td>1540.0</td>\n",
       "      <td>...</td>\n",
       "      <td>380.0</td>\n",
       "      <td>389.0</td>\n",
       "      <td>381.0</td>\n",
       "      <td>524.0</td>\n",
       "      <td>1222.0</td>\n",
       "      <td>378.0</td>\n",
       "      <td>75.0</td>\n",
       "      <td>237.0</td>\n",
       "      <td>2443.0</td>\n",
       "      <td>3112.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAB_P</th>\n",
       "      <td>704.0</td>\n",
       "      <td>203.0</td>\n",
       "      <td>785.0</td>\n",
       "      <td>4871.0</td>\n",
       "      <td>299.0</td>\n",
       "      <td>397.0</td>\n",
       "      <td>1078.0</td>\n",
       "      <td>3087.0</td>\n",
       "      <td>284.0</td>\n",
       "      <td>331.0</td>\n",
       "      <td>...</td>\n",
       "      <td>384.0</td>\n",
       "      <td>394.0</td>\n",
       "      <td>403.0</td>\n",
       "      <td>306.0</td>\n",
       "      <td>1102.0</td>\n",
       "      <td>270.0</td>\n",
       "      <td>185.0</td>\n",
       "      <td>268.0</td>\n",
       "      <td>1981.0</td>\n",
       "      <td>4670.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAC_P</th>\n",
       "      <td>471.0</td>\n",
       "      <td>117.0</td>\n",
       "      <td>431.0</td>\n",
       "      <td>3418.0</td>\n",
       "      <td>421.0</td>\n",
       "      <td>360.0</td>\n",
       "      <td>788.0</td>\n",
       "      <td>1421.0</td>\n",
       "      <td>68.0</td>\n",
       "      <td>510.0</td>\n",
       "      <td>...</td>\n",
       "      <td>351.0</td>\n",
       "      <td>220.0</td>\n",
       "      <td>244.0</td>\n",
       "      <td>239.0</td>\n",
       "      <td>1255.0</td>\n",
       "      <td>334.0</td>\n",
       "      <td>102.0</td>\n",
       "      <td>324.0</td>\n",
       "      <td>2225.0</td>\n",
       "      <td>2421.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAF_P</th>\n",
       "      <td>517.0</td>\n",
       "      <td>220.0</td>\n",
       "      <td>289.0</td>\n",
       "      <td>3521.0</td>\n",
       "      <td>206.0</td>\n",
       "      <td>326.0</td>\n",
       "      <td>555.0</td>\n",
       "      <td>1837.0</td>\n",
       "      <td>104.0</td>\n",
       "      <td>1460.0</td>\n",
       "      <td>...</td>\n",
       "      <td>368.0</td>\n",
       "      <td>285.0</td>\n",
       "      <td>387.0</td>\n",
       "      <td>476.0</td>\n",
       "      <td>886.0</td>\n",
       "      <td>262.0</td>\n",
       "      <td>62.0</td>\n",
       "      <td>177.0</td>\n",
       "      <td>1100.0</td>\n",
       "      <td>2649.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAG_P</th>\n",
       "      <td>655.0</td>\n",
       "      <td>259.0</td>\n",
       "      <td>639.0</td>\n",
       "      <td>5450.0</td>\n",
       "      <td>305.0</td>\n",
       "      <td>320.0</td>\n",
       "      <td>747.0</td>\n",
       "      <td>874.0</td>\n",
       "      <td>126.0</td>\n",
       "      <td>526.0</td>\n",
       "      <td>...</td>\n",
       "      <td>159.0</td>\n",
       "      <td>97.0</td>\n",
       "      <td>331.0</td>\n",
       "      <td>263.0</td>\n",
       "      <td>1164.0</td>\n",
       "      <td>318.0</td>\n",
       "      <td>133.0</td>\n",
       "      <td>167.0</td>\n",
       "      <td>925.0</td>\n",
       "      <td>3495.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GLSS-SM-R088_P</th>\n",
       "      <td>2696.0</td>\n",
       "      <td>270.0</td>\n",
       "      <td>1085.0</td>\n",
       "      <td>273.0</td>\n",
       "      <td>705.0</td>\n",
       "      <td>680.0</td>\n",
       "      <td>1116.0</td>\n",
       "      <td>1273.0</td>\n",
       "      <td>977.0</td>\n",
       "      <td>558.0</td>\n",
       "      <td>...</td>\n",
       "      <td>517.0</td>\n",
       "      <td>127.0</td>\n",
       "      <td>318.0</td>\n",
       "      <td>457.0</td>\n",
       "      <td>799.0</td>\n",
       "      <td>664.0</td>\n",
       "      <td>673.0</td>\n",
       "      <td>1799.0</td>\n",
       "      <td>630.0</td>\n",
       "      <td>8508.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GLSS-SM-R093_R</th>\n",
       "      <td>1587.0</td>\n",
       "      <td>511.0</td>\n",
       "      <td>1655.0</td>\n",
       "      <td>651.0</td>\n",
       "      <td>1571.0</td>\n",
       "      <td>1209.0</td>\n",
       "      <td>2911.0</td>\n",
       "      <td>569.0</td>\n",
       "      <td>456.0</td>\n",
       "      <td>1004.0</td>\n",
       "      <td>...</td>\n",
       "      <td>360.0</td>\n",
       "      <td>201.0</td>\n",
       "      <td>451.0</td>\n",
       "      <td>236.0</td>\n",
       "      <td>1201.0</td>\n",
       "      <td>539.0</td>\n",
       "      <td>563.0</td>\n",
       "      <td>1147.0</td>\n",
       "      <td>543.0</td>\n",
       "      <td>1919.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GLSS-SM-R093_P</th>\n",
       "      <td>2020.0</td>\n",
       "      <td>782.0</td>\n",
       "      <td>2248.0</td>\n",
       "      <td>644.0</td>\n",
       "      <td>1673.0</td>\n",
       "      <td>1132.0</td>\n",
       "      <td>2502.0</td>\n",
       "      <td>1099.0</td>\n",
       "      <td>61.0</td>\n",
       "      <td>1286.0</td>\n",
       "      <td>...</td>\n",
       "      <td>593.0</td>\n",
       "      <td>371.0</td>\n",
       "      <td>830.0</td>\n",
       "      <td>236.0</td>\n",
       "      <td>1433.0</td>\n",
       "      <td>606.0</td>\n",
       "      <td>741.0</td>\n",
       "      <td>1136.0</td>\n",
       "      <td>1164.0</td>\n",
       "      <td>3447.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-06-0125_R</th>\n",
       "      <td>1084.0</td>\n",
       "      <td>275.0</td>\n",
       "      <td>1569.0</td>\n",
       "      <td>812.0</td>\n",
       "      <td>1535.0</td>\n",
       "      <td>859.0</td>\n",
       "      <td>2860.0</td>\n",
       "      <td>900.0</td>\n",
       "      <td>723.0</td>\n",
       "      <td>870.0</td>\n",
       "      <td>...</td>\n",
       "      <td>611.0</td>\n",
       "      <td>332.0</td>\n",
       "      <td>404.0</td>\n",
       "      <td>511.0</td>\n",
       "      <td>1272.0</td>\n",
       "      <td>744.0</td>\n",
       "      <td>328.0</td>\n",
       "      <td>180.0</td>\n",
       "      <td>1053.0</td>\n",
       "      <td>11983.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-06-0125_P</th>\n",
       "      <td>2602.0</td>\n",
       "      <td>918.0</td>\n",
       "      <td>2147.0</td>\n",
       "      <td>427.0</td>\n",
       "      <td>3084.0</td>\n",
       "      <td>1477.0</td>\n",
       "      <td>3777.0</td>\n",
       "      <td>1743.0</td>\n",
       "      <td>454.0</td>\n",
       "      <td>1431.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1037.0</td>\n",
       "      <td>511.0</td>\n",
       "      <td>1197.0</td>\n",
       "      <td>878.0</td>\n",
       "      <td>1832.0</td>\n",
       "      <td>1027.0</td>\n",
       "      <td>1197.0</td>\n",
       "      <td>1092.0</td>\n",
       "      <td>2158.0</td>\n",
       "      <td>35720.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>608 rows × 5829 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "feature_name      AAAS  AADAT   AAGAB    AAK1    AAR2  AARSD1  AASDHPPT  \\\n",
       "JAN_P            543.0  160.0   518.0  4397.0   458.0   429.0     706.0   \n",
       "AAB_P            704.0  203.0   785.0  4871.0   299.0   397.0    1078.0   \n",
       "AAC_P            471.0  117.0   431.0  3418.0   421.0   360.0     788.0   \n",
       "AAF_P            517.0  220.0   289.0  3521.0   206.0   326.0     555.0   \n",
       "AAG_P            655.0  259.0   639.0  5450.0   305.0   320.0     747.0   \n",
       "...                ...    ...     ...     ...     ...     ...       ...   \n",
       "GLSS-SM-R088_P  2696.0  270.0  1085.0   273.0   705.0   680.0    1116.0   \n",
       "GLSS-SM-R093_R  1587.0  511.0  1655.0   651.0  1571.0  1209.0    2911.0   \n",
       "GLSS-SM-R093_P  2020.0  782.0  2248.0   644.0  1673.0  1132.0    2502.0   \n",
       "TCGA-06-0125_R  1084.0  275.0  1569.0   812.0  1535.0   859.0    2860.0   \n",
       "TCGA-06-0125_P  2602.0  918.0  2147.0   427.0  3084.0  1477.0    3777.0   \n",
       "\n",
       "feature_name      AASS   AATK   ABCA5  ...  ZSCAN21  ZSCAN5A  ZSCAN9  ZSWIM4  \\\n",
       "JAN_P           1541.0  717.0  1540.0  ...    380.0    389.0   381.0   524.0   \n",
       "AAB_P           3087.0  284.0   331.0  ...    384.0    394.0   403.0   306.0   \n",
       "AAC_P           1421.0   68.0   510.0  ...    351.0    220.0   244.0   239.0   \n",
       "AAF_P           1837.0  104.0  1460.0  ...    368.0    285.0   387.0   476.0   \n",
       "AAG_P            874.0  126.0   526.0  ...    159.0     97.0   331.0   263.0   \n",
       "...                ...    ...     ...  ...      ...      ...     ...     ...   \n",
       "GLSS-SM-R088_P  1273.0  977.0   558.0  ...    517.0    127.0   318.0   457.0   \n",
       "GLSS-SM-R093_R   569.0  456.0  1004.0  ...    360.0    201.0   451.0   236.0   \n",
       "GLSS-SM-R093_P  1099.0   61.0  1286.0  ...    593.0    371.0   830.0   236.0   \n",
       "TCGA-06-0125_R   900.0  723.0   870.0  ...    611.0    332.0   404.0   511.0   \n",
       "TCGA-06-0125_P  1743.0  454.0  1431.0  ...   1037.0    511.0  1197.0   878.0   \n",
       "\n",
       "feature_name    ZSWIM6    ZW10  ZWILCH   ZWINT    ZXDC      ZYX  \n",
       "JAN_P           1222.0   378.0    75.0   237.0  2443.0   3112.0  \n",
       "AAB_P           1102.0   270.0   185.0   268.0  1981.0   4670.0  \n",
       "AAC_P           1255.0   334.0   102.0   324.0  2225.0   2421.0  \n",
       "AAF_P            886.0   262.0    62.0   177.0  1100.0   2649.0  \n",
       "AAG_P           1164.0   318.0   133.0   167.0   925.0   3495.0  \n",
       "...                ...     ...     ...     ...     ...      ...  \n",
       "GLSS-SM-R088_P   799.0   664.0   673.0  1799.0   630.0   8508.0  \n",
       "GLSS-SM-R093_R  1201.0   539.0   563.0  1147.0   543.0   1919.0  \n",
       "GLSS-SM-R093_P  1433.0   606.0   741.0  1136.0  1164.0   3447.0  \n",
       "TCGA-06-0125_R  1272.0   744.0   328.0   180.0  1053.0  11983.0  \n",
       "TCGA-06-0125_P  1832.0  1027.0  1197.0  1092.0  2158.0  35720.0  \n",
       "\n",
       "[608 rows x 5829 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "def pyCheckData(df):\n",
    "    \n",
    "    errors = []\n",
    "    warnings = []\n",
    "    \n",
    "    # Import required genes\n",
    "    gene_lengths = pd.read_csv(\"../data/GBMPurity_genes.csv\")\n",
    "    genes = gene_lengths['feature_name']\n",
    "    \n",
    "    # Check Dimensions\n",
    "    if df.shape[0] < 1:\n",
    "        errors.append(\"We didn't detect any genes.\")\n",
    "        return errors, warnings, None\n",
    "        \n",
    "    if df.shape[1] <= 1:\n",
    "        errors.append(\"We didn't detect any samples.\")\n",
    "        return errors, warnings, None\n",
    "    \n",
    "    # Missing values\n",
    "    if df.isnull().values.any():\n",
    "        warnings.append(f\"We found {df.isnull().values.sum()} missing values. These will be converted to 0.\")\n",
    "        df = df.fillna(0)\n",
    "    \n",
    "    # Check for duplicate genes\n",
    "    input_genes = df.iloc[:,0]\n",
    "    duplicate_genes = input_genes[input_genes.duplicated()].unique()\n",
    "    if len(duplicate_genes) > 0:\n",
    "        warnings.append(f'We found {len(duplicate_genes)} duplicate genes. Counts for these genes will be summed for each sample.')\n",
    "    \n",
    "    data = df.set_index(df.columns[0])\n",
    "    data = data.groupby(data.index).sum()\n",
    "    \n",
    "    # Check appropriate genes\n",
    "    overlap = set(input_genes).intersection(set(genes)) \n",
    "    if len(overlap) == 0:\n",
    "        errors.append(\"We didn't find any required genes. Are the provided genes in the HGNC format e.g. CD47?\")\n",
    "        return errors, warnings, None\n",
    "    else:\n",
    "        p_overlap = len(overlap)/len(genes)\n",
    "        if p_overlap < 0.8:\n",
    "            errors.append(f\"We only found {p_overlap * 100}% of the required genes. Purity estimates will be unreliable under 80%.\")\n",
    "            return errors, warnings, None\n",
    "        elif p_overlap < 0.99:\n",
    "            warnings.append(f\"We found {int(p_overlap * 100)}% of the required genes. Note that GBMPurity tends to underestimate the tumour purity with more missing genes.\")\n",
    "    \n",
    "\n",
    "    # Check correct data\n",
    "    # Non-numeric\n",
    "    non_numeric = data.apply(lambda s: pd.to_numeric(s, errors='coerce').notnull().all()).all()\n",
    "    if not non_numeric:\n",
    "        errors.append(\"All gene expression values must be numeric.\")\n",
    "        return errors, warnings, None\n",
    "\n",
    "    # Negative values\n",
    "    if (data.values < 0).any():\n",
    "        errors.append(\"Gene expression values must be non-negative. Data should be uploaded as raw counts (without batch correction).\")\n",
    "        return errors, warnings, None\n",
    "    \n",
    "    # Process data to return\n",
    "    data = data.T\n",
    "    data = data.reindex(columns=genes, fill_value=0)\n",
    "\n",
    "    # Check \n",
    "    return errors, warnings, data\n",
    "\n",
    "\n",
    "e, w, data = pyCheckData(data)\n",
    "\n",
    "display(data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample</th>\n",
       "      <th>Purity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>JAN_P</td>\n",
       "      <td>0.544</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AAB_P</td>\n",
       "      <td>0.609</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AAC_P</td>\n",
       "      <td>0.823</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AAF_P</td>\n",
       "      <td>0.705</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AAG_P</td>\n",
       "      <td>0.208</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>603</th>\n",
       "      <td>GLSS-SM-R088_P</td>\n",
       "      <td>0.456</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>604</th>\n",
       "      <td>GLSS-SM-R093_R</td>\n",
       "      <td>0.317</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>605</th>\n",
       "      <td>GLSS-SM-R093_P</td>\n",
       "      <td>0.670</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>606</th>\n",
       "      <td>TCGA-06-0125_R</td>\n",
       "      <td>0.719</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>607</th>\n",
       "      <td>TCGA-06-0125_P</td>\n",
       "      <td>0.784</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>608 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             Sample  Purity\n",
       "0             JAN_P   0.544\n",
       "1             AAB_P   0.609\n",
       "2             AAC_P   0.823\n",
       "3             AAF_P   0.705\n",
       "4             AAG_P   0.208\n",
       "..              ...     ...\n",
       "603  GLSS-SM-R088_P   0.456\n",
       "604  GLSS-SM-R093_R   0.317\n",
       "605  GLSS-SM-R093_P   0.670\n",
       "606  TCGA-06-0125_R   0.719\n",
       "607  TCGA-06-0125_P   0.784\n",
       "\n",
       "[608 rows x 2 columns]"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def GBMPurity(data):\n",
    "    \n",
    "    # Import gene lengths\n",
    "    gene_lengths = pd.read_csv(\"../data/GBMPurity_genes.csv\")\n",
    "    lengths = gene_lengths['feature_length'].values\n",
    "\n",
    "    # Transform input data\n",
    "    X = np.log2(tpm(data.values, lengths) + 1)\n",
    "    \n",
    "    # Import model\n",
    "    model = torch.load(\"./model/GBMPurity.pt\")\n",
    "    model.eval()\n",
    "    \n",
    "    # Input to GBMPurity\n",
    "    y_pred = model(torch.tensor(X).float()).detach().numpy().flatten().clip(0, 1)\n",
    "    \n",
    "    samples = data.index.values\n",
    "    results = pd.DataFrame({'Sample':samples, 'Purity':y_pred.round(3)})\n",
    "    return results\n",
    "\n",
    "\n",
    "GBMPurity(data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Unit testing\n",
    "valid_data = pd.DataFrame({\n",
    "    'GeneName': ['AAAS', 'AADAT', 'AAGAB', 'AAK1', 'AAR2'],\n",
    "    'Sample1': [543, 160, 518, 4397, 458],\n",
    "    'Sample2': [704, 203, 785, 4871, 299]\n",
    "})\n",
    "\n",
    "no_genes_data = pd.DataFrame({\n",
    "    'GeneName': [],\n",
    "    'Sample1': [],\n",
    "    'Sample2': []\n",
    "})\n",
    "\n",
    "no_samples_data = pd.DataFrame({\n",
    "    'GeneName': ['AAAS', 'AADAT', 'AAGAB', 'AAK1', 'AAR2']\n",
    "})\n",
    "\n",
    "duplicate_genes_data = pd.DataFrame({\n",
    "    'GeneName': ['AAAS', 'AADAT', 'AADAT', 'AAGAB', 'AAK1'],\n",
    "    'Sample1': [543, 160, 160, 518, 4397],\n",
    "    'Sample2': [704, 203, 203, 785, 4871]\n",
    "})\n",
    "\n",
    "no_required_genes_data = pd.DataFrame({\n",
    "    'GeneName': ['UNKNOWN1', 'UNKNOWN2', 'UNKNOWN3'],\n",
    "    'Sample1': [543, 160, 518],\n",
    "    'Sample2': [704, 203, 785]\n",
    "})\n",
    "\n",
    "missing_values_data = pd.DataFrame({\n",
    "    'GeneName': ['AAAS', 'AADAT', 'AAGAB', 'AAK1', 'AAR2'],\n",
    "    'Sample1': [543, 160, None, 4397, 458],\n",
    "    'Sample2': [704, None, 785, 4871, 299]\n",
    "})\n",
    "\n",
    "non_numeric_data = pd.DataFrame({\n",
    "    'GeneName': ['AAAS', 'AADAT', 'AAGAB', 'AAK1', 'AAR2'],\n",
    "    'Sample1': [543, 'one-sixty', 518, 4397, 458],\n",
    "    'Sample2': [704, 203, 'seven-eighty-five', 4871, 299]\n",
    "})\n",
    "\n",
    "negative_values_data = pd.DataFrame({\n",
    "    'GeneName': ['AAAS', 'AADAT', 'AAGAB', 'AAK1', 'AAR2'],\n",
    "    'Sample1': [543, -160, 518, 4397, 458],\n",
    "    'Sample2': [704, 203, 785, -4871, 299]\n",
    "})\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[\"We didn't detect any samples.\"]\n",
      "[]\n"
     ]
    }
   ],
   "source": [
    "e, w, test_data = pyCheckData(no_samples_data)\n",
    "print(e)\n",
    "print(w)\n",
    "\n",
    "if len(e) == 0:\n",
    "    purities = GBMPurity(test_data)\n",
    "    display(purities)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Test Data Preparation\n",
    "no_genes_df = pd.DataFrame(columns=[\"GeneName\", \"Sample1\", \"Sample2\"])\n",
    "no_genes_df.to_csv(\"../data/test/no_genes.csv\", index=False)\n",
    "\n",
    "no_samples_df = pd.DataFrame({\"GeneName\": [\"Gene1\", \"Gene2\"]})\n",
    "no_samples_df.to_csv(\"../data/test/no_samples.csv\", index=False)\n",
    "\n",
    "missing_values_df = pd.read_csv(\"../data/GBMPurity-test-input.csv\")\n",
    "missing_values_df.iloc[0, 1] = np.nan\n",
    "missing_values_df.to_csv(\"../data/test/missing_values.csv\", index=False)\n",
    "\n",
    "duplicate_genes_df = pd.read_csv(\"../data/GBMPurity-test-input.csv\")\n",
    "duplicate_genes_df = duplicate_genes_df.append(duplicate_genes_df.iloc[0])\n",
    "duplicate_genes_df.to_csv(\"../data/test/duplicate_genes.csv\", index=False)\n",
    "\n",
    "no_required_genes_df = pd.read_csv(\"../data/GBMPurity-test-input.csv\")\n",
    "no_required_genes_df[\"GeneName\"] = [\"NonHGNC1\", \"NonHGNC2\"]\n",
    "no_required_genes_df.to_csv(\"../data/test/no_required_genes.csv\", index=False)\n",
    "\n",
    "non_numeric_df = pd.read_csv(\"../data/GBMPurity-test-input.csv\")\n",
    "non_numeric_df.iloc[0, 1] = \"non-numeric\"\n",
    "non_numeric_df.to_csv(\"../data/test/non_numeric_values.csv\", index=False)\n",
    "\n",
    "negative_values_df = pd.read_csv(\"../data/GBMPurity-test-input.csv\")\n",
    "negative_values_df.iloc[0, 1] = -1\n",
    "negative_values_df.to_csv(\"../data/test/negative_values.csv\", index=False)\n",
    "\n",
    "less_than_80_percent_genes_df = pd.read_csv(\"../data/GBMPurity-test-input.csv\")\n",
    "gene_lengths = pd.read_csv(\"data/GBMPurity_genes.csv\")\n",
    "required_genes = gene_lengths['feature_name']\n",
    "less_than_80_percent_genes = required_genes[:int(0.79 * len(required_genes))]\n",
    "less_than_80_percent_genes_df = less_than_80_percent_genes_df[less_than_80_percent_genes_df['GeneName'].isin(less_than_80_percent_genes)]\n",
    "less_than_80_percent_genes_df.to_csv(\"../data/test/less_than_80_percent_genes.csv\", index=False)\n",
    "\n",
    "between_80_and_99_percent_genes_df = pd.read_csv(\"../data/GBMPurity-test-input.csv\")\n",
    "between_80_and_99_percent_genes = required_genes[:int(0.95 * len(required_genes))]\n",
    "between_80_and_99_percent_genes_df = between_80_and_99_percent_genes_df[between_80_and_99_percent_genes_df['GeneName'].isin(between_80_and_99_percent_genes)]\n",
    "between_80_and_99_percent_genes_df.to_csv(\"../data/test/between_80_and_99_percent_genes.csv\", index=False)\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "GBMPurity",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
